File: performance_test/python/sieve.py
import time

def sieve(limit):
    # Using bytearray for efficiency in Python
    primes = bytearray([1]) * (limit + 1)
    primes[0] = primes[1] = 0
    
    for p in range(2, int(limit**0.5) + 1):
        if primes[p]:
            for i in range(p * p, limit + 1, p):
                primes[i] = 0
    
    return sum(primes)

if __name__ == "__main__":
    limit = 100_000_000
    print(f"Python: Starting sieve up to {limit:,}...")
    start = time.time()
    count = sieve(limit)
    end = time.time()
    print(f"Python: Found {count} primes.")
    print(f"Python: Time taken: {end - start:.4f} seconds")

File: performance_test/php/sieve.php
<?php
ini_set('memory_limit', '2G');

function sieve($limit) {
    // SplFixedArray is more memory efficient than standard arrays
    $primes = new SplFixedArray($limit + 1);
    for ($i = 0; $i <= $limit; $i++) {
        $primes[$i] = true;
    }
    $primes[0] = $primes[1] = false;

    for ($p = 2; $p * $p <= $limit; $p++) {
        if ($primes[$p]) {
            for ($i = $p * $p; $i <= $limit; $i += $p) {
                $primes[$i] = false;
            }
        }
    }

    $count = 0;
    foreach ($primes as $isPrime) {
        if ($isPrime) $count++;
    }
    return $count;
}

$limit = 100000000;
echo "PHP: Starting sieve up to " . number_format($limit) . "...\n";
$start = microtime(true);
$count = sieve($limit);
$end = microtime(true);
echo "PHP: Found $count primes.\n";
echo "PHP: Time taken: " . number_format($end - $start, 4) . " seconds\n";
